blockchain/src/pretty.rs

```diff
@@ -0,0 +1,405 @@
+use std;
+use std::fmt;
+use std::string::String;
+
+use ansi_term::Colour;
+
+use block;
+use genesis;
+use normal;
+use types;
+use wallet_crypto;
+
+static DISPLAY_INDENT_SIZE: usize = 4; // spaces
+static DISPLAY_INDENT_LEVEL: usize = 0; // beginning starts at zero
+static DISPLAY_USE_COLOR: bool = false; // no color for display implementations
+
+type Assoc = (String, Val);
+
+// core pretty-printer type which facilitates homogenous ordered key-value pairs
+pub enum Val {
+    Single(Box<fmt::Display>),
+    Pairs(Option<Colour>, Vec<Assoc>),
+}
+
+// XXX: this class is related to "into" somehow
+pub trait Pretty {
+    fn to_pretty(&self) -> Val;
+}
+
+fn longest_key_length(pairs: &Vec<Assoc>) -> usize {
+    pairs
+        .iter()
+        .fold(0, |longest, (key, _)| std::cmp::max(longest, key.len()))
+}
+
+fn fmt_key(
+    key: &String,
+    f: &mut fmt::Formatter,
+    indent_size: usize,
+    indent_level: usize,
+    col: Option<Colour>,
+    key_width: usize,
+) -> fmt::Result {
+    // XXX: measuring width is broken for ansi strings, so do alignment before painting
+    let aligned_key = format!("{:<kw$}", key, kw = key_width,);
+    let painted_key = match col {
+        Some(color) => color.paint(aligned_key),
+        None => aligned_key.into(),
+    };
+    write!(
+        f,
+        "{:>iw$}- {}:",
+        "",
+        painted_key,
+        iw = indent_size * indent_level,
+    )
+}
+
+fn fmt_val(
+    val: &Val,
+    f: &mut fmt::Formatter,
+    indent_size: usize,
+    indent_level: usize,
+    use_color: bool,
+) -> fmt::Result {
+    match val {
+        // write inline
+        Val::Single(_) => {
+            write!(f, " ")?;
+            fmt_pretty(val, f, indent_size, indent_level, use_color)?;
+            write!(f, "\n")
+        }
+        // write on the next line
+        Val::Pairs(_, _) => {
+            write!(f, "\n")?;
+            fmt_pretty(val, f, indent_size, indent_level, use_color)
+        }
+    }
+    // XXX: DRY up the duplicate calls to `fmt_pretty`?
+}
+
+fn fmt_pretty(
+    p: &Val,
+    f: &mut fmt::Formatter,
+    indent_size: usize,
+    indent_level: usize,
+    use_color: bool,
+) -> fmt::Result {
+    match p {
+        // format pretty-val as a terminal
+        Val::Single(display) => write!(f, "{}", display),
+        // format pretty-val as a set of  key-vals
+        Val::Pairs(color, pairs) => {
+            let key_width = longest_key_length(pairs);
+            pairs
+                .iter()
+                .fold(Ok(()), |prev_result, (key, val)| match prev_result {
+                    // return an error
+                    err @ Err(_) => err,
```

Comment:
no, just inexperience with rust.. how do you exit the iter on error? woudl using `and_then` as @NicolasDP suggested do it like in the monad world?

Reply:
If you encounter an error, I assume that you actually want to return from the function with an error, which you can achieve by using `return Err` instead of `err`.

That's what the `?` notation does for you: `a?` is equivalent to
```rust
match a {
  err @ Err(_) => return err, -- this returns immediately from the function
  Ok(a') => a'
}
```
It's quite convenient, if you don't intend to manage the error in the same function.

Reply:
I think `=> err` returning from the closure on line 97.

Would `=> return err,` be any different? What are you saying itnwould return from?

Reply:
I thought so, but I was wrong. So, yeah, using `?` here would not be different from using `and_then`.